Keymapper phase 2 enhancements

rexagod opened the pull request:
Fixes #239 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

rexagod commented:
## Next up

- "along the way lets figure out how to make keymapper_options and keymapper more convenient, perhaps merge them into one option."

- "Update docs"

/cc @jywarren @sashadev-sky

rexagod commented:
Okay so how I've merged `keymapper_position` and `keymapper` is in a `position` param that will render the keymapper only when there's a non-null value passed to it.
```js
 keymapper = new L.DistortableImage.Keymapper(map,img., {
        position: '' //keymapper not rendered
      });   
```

rexagod commented:
Open to other ideas regarding merging both these params, else I guess maybe we can proceed with this?

rexagod commented:
In response to #258
 
![Untitled](https://user-images.githubusercontent.com/33557095/58373915-f8d0f380-7f52-11e9-8664-f10105a0a292.png)

cc @jywarren

rexagod commented:
@sashadev-sky I've noted your suggestion and will chime in a bit after I wrap up the CLI tool if it's not a problem? Thanks!

jywarren commented:
Oh this looks good! Thanks! If this is getting ready for a merge and
release, let me know but if it'll be another little while, we could
consider just using an "x" button, what do you think?

On Sat, May 25, 2019, 3:42 PM Pranshu Srivastava <notifications@github.com>
wrote:

> @sashadev-sky <https://github.com/sashadev-sky> I've noted your
> suggestion and will chime in a bit after I wrap up the CLI tool if it's not
> a problem? Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/Leaflet.DistortableImage/pull/243?email_source=notifications&email_token=AAAF6JYRME2WWWYBOJQSE4TPXGI6TA5CNFSM4HKUZDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWHX7CI#issuecomment-495943561>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAF6J6FM2KUXYM6HT4D65TPXGI6TANCNFSM4HKUZDDQ>
> .
>

rexagod commented:
Okay, no problem, and just to be clear, where should I place this 'x' button? Do you think that maybe we should use a fontawesome icon inside the button above (replace 'disable keymapper' with 'x' icon) or is there any particular position that would make more sense?

jywarren commented:
Yeah, a fontawesome icon sounds... awesome! And yeah, I guess upper corner
would be perfect. Thanks!

On Sat, May 25, 2019, 3:55 PM Pranshu Srivastava <notifications@github.com>
wrote:

> Okay, no problem, and just to be clear, where should I place this 'x'
> button? Do you think that maybe we should use a fontawesome icon inside the
> button above (replace 'disable keymapper' with 'x' icon) or is there any
> particular position that would make more sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/Leaflet.DistortableImage/pull/243?email_source=notifications&email_token=AAAF6J46DKWMBF24OQNCSXDPXGKQ7A5CNFSM4HKUZDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWHYFPI#issuecomment-495944381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAF6J4NWCTIBZTMXXI3NZTPXGKQ7ANCNFSM4HKUZDDQ>
> .
>

rexagod commented:
@jywarren How does this look?

![Capture](https://user-images.githubusercontent.com/33557095/58377518-69523180-7fa0-11e9-84b5-02386eafe197.PNG)

sashadev-sky commented:
@jywarren @rexagod do you think ability to hide keymapper should be addressed in a separate PR? And we can use a toggled icon for minimizing and expanding it? Because deleting is a permanent action and that might confuse people

also still need to review the implementation for this with @rexagod!

rexagod commented:
> Would love some insight here because I could be misunderstanding the point! Also I think `@jywarren` initially suggested used L.handler. Was there a specific reason for that?

Hmm. I guess his response here should definitely pave the way for this PR. Pinging @jywarren.

> I was also unsure about its functionality because I found that after the keymapper is initially enabled, running keymapper.disable() does not remove it from the map, as img.editing.disable() removes the markers and toolbars from an image.

@sashadev-sky `disable()` seems to work, or am I running a different use case scenario?

![test](https://user-images.githubusercontent.com/33557095/58592544-52545d80-8286-11e9-9fd5-614884abe05f.gif)

jywarren commented:
I'm fine with splitting things out, but had asked @rexagod to prioritize
the "close" button because it's causing some difficulties in live
MapKnitter.org. Thank you!

As to handler... gosh, i feel like it was a million years ago. I honestly
don't remember, but if you can pull the link to the comment i can try to
catch up on it?

On Wed, May 29, 2019 at 5:27 PM Pranshu Srivastava <notifications@github.com>
wrote:

> Would love some insight here because I could be misunderstanding the
> point! Also I think @jywarren initially suggested used L.handler. Was
> there a specific reason for that?
>
> Hmm. I guess his response here should definitely pave the way for this PR.
> Pinging @jywarren <https://github.com/jywarren>.
>
> I was also unsure about its functionality because I found that after the
> keymapper is initially enabled, running keymapper.disable() does not remove
> it from the map, as img.editing.disable() removes the markers and toolbars
> from an image.
>
> @sashadev-sky <https://github.com/sashadev-sky> disable() seems to work,
> or am I running a different use case scenario?
>
> [image: test]
> <https://user-images.githubusercontent.com/33557095/58592544-52545d80-8286-11e9-9fd5-614884abe05f.gif>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/Leaflet.DistortableImage/pull/243?email_source=notifications&email_token=AAAF6J3BO5UE77LVWDPFQYDPX3YIRA5CNFSM4HKUZDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWQWA5A#issuecomment-497115252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAF6J4IOLFZKDHDJT3BYNTPX3YIRANCNFSM4HKUZDDQ>
> .
>

sashadev-sky commented:
@jywarren https://github.com/publiclab/Leaflet.DistortableImage/pull/212#issuecomment-488114870 I think it might have been on accident you suggested but I could be wrong! 

@rexagod hmm I think what I was referring to was that if you try to use disable() later on, like in the console perhaps, it doesn't disable the keymapper like it would with the handles. 

If there isn't an identifiable benefit to making it in a handler class that then call the control class to create it, should we just refactor to make it control class like in my suggestion above?

jywarren commented:
I just turned off the keymapper in MapKnitter briefly until we get the "many keymappers appearing" and "ability to close keymapper" resolved: https://github.com/publiclab/Leaflet.DistortableImage/pull/243

Thanks!

I also thought we ought to make it not /ever/ open a 2nd keymapper, does that make sense? 

Awesome!

jywarren commented:
That should relieve pressure on this PR so we can take a bit more time to get it right. Thanks, all!

jywarren commented:
![image](https://user-images.githubusercontent.com/24359/58656993-5797c380-82eb-11e9-982c-5a82ab6eea8d.png)

jywarren commented:
> Also I think @jywarren initially suggested used L.handler. Was there a specific reason for that?

Aha! So, in https://github.com/publiclab/Leaflet.DistortableImage/pull/212#issuecomment-488114870, I meant that we could give it its own namespace, /like/ the `L.Handler` example, but didn't mean that it should actually be named `Handler`. So, `L.Keymapper` maybe, or if it's very LDI specific, maybe `L.DistortableImageKeymapper` -- or even `L.DistortableImage.Keymapper`... although that then isn't as modular again. Make sense?

jywarren commented on dist/leaflet.distortableimage.js:
```diff
@@ -1133,14 +1133,27 @@ L.Map.addInitHook('addHandler', 'boxSelector', L.Map.BoxSelectHandle);
-L.DistortableImage.Keymapper = L.Control.extend({
-    initialize: function(options) {
-        L.Control.prototype.initialize.call(this, options);
+L.DistortableImage.Keymapper = L.Handler.extend({
```
So here, yeah - sorry, we don't need this to extend `Handler` - that was just a fragment of the code i was quoting. I mainly meant just putting this all in `L.DistortableImage.Keymapper` - which looks great!

jywarren commented:
Sorry, clarifying further re: Handler, i put an inline comment at https://github.com/publiclab/Leaflet.DistortableImage/pull/243#discussion_r289128088

rexagod commented:
>And we can use a toggled icon for minimizing and expanding it? Because deleting is a permanent action and that might confuse people

> I think what I was referring to was that if you try to use disable() later on, like in the console perhaps, it doesn't disable the keymapper like it would with the handles.

>I also thought we ought to make it not /ever/ open a 2nd keymapper, does that make sense?

![test](https://user-images.githubusercontent.com/33557095/58753571-b6715e80-84de-11e9-8e99-b72771a1d7e1.gif)

@jywarren @sashadev-sky Okay so I've added the aforementioned changes to this, while holding on to the Handler -> Control one, for the following reasons.

- No Hooks.
- No enable/enabled methods.
- Leaflet's doc on L.Control: "L.Control is a base class for implementing **map controls**. Handles positioning. All other controls extend from this class." Now in case of keymapper, I really don't think we need need to "control" the map in any way (like the zoom buttons at the topleft), just the keymapper itself. As far as the positioning is concerned, we can already do that by passing the position parameter at instantiation.
Therefore, I didn't see any actual benefit from the refactoring, so I thought of clearing this out first. But I can be wrong, and thus ask your views on the same.

Thanks!!

rexagod commented:
Also, I'm currently using Handler wrapper class just for the above mentioned functionality it provides, while the actual keymapper is implemented in `Control` class.

```js
var keymapper = L.control({position: this._position});
        keymapper.onAdd = function() {
            var el_wrapper = L.DomUtil.create("div", "l-container");
            el_wrapper.setAttribute('id', 'keymapper-wrapper');
            el_wrapper.innerHTML = "<table><tbody>" +
            "<tr><td><center><span id='keymapper-heading'>Keymappings</span></center></td></tr>" +
            "<tr><td id='keymapper-hr'><hr></td></tr>" +
            "<tr><td><kbd>j</kbd>, <kbd>k</kbd>: <span>Send up / down (stacking order)</span></td></tr>" +
            "<tr><td><kbd>l</kbd>: <span>Lock</span></td></tr>" +
            "<tr><td><kbd>o</kbd>: <span>Outline</span></td></tr>" +
            "<tr><td><kbd>s</kbd>: <span>Scale</span></td></tr>" +
            "<tr><td><kbd>t</kbd>: <span>Transparency</span></td></tr>" +
            "<tr><td><kbd>d</kbd> , <kbd>r</kbd>: <span>RotateScale</span> </td></tr>" +
            "<tr><td><kbd>esc</kbd>: <span>Deselect All</span></td></tr>" +
            "<tr><td><kbd>delete</kbd> , <kbd>backspace</kbd>: <span>Delete</span></td></tr>" +
            "<tr><td><kbd>caps</kbd>: <span>Rotate</span></td></tr>" +
            "<tr><td><center><button id='close-keymapper-button'>" + 
            "<kbd id='close-keymapper-kbd'>" + 
            "<i class='fa fa-times-circle' aria-hidden='true'></i>" + 
            "</kbd></button></center></td></tr>" +
            "</tbody></table>";
            return el_wrapper;
        };
        keymapper.addTo(this._map);
```

sashadev-sky commented:
@rexagod I am going to pull this and rebase you because you've been out of the repo for a bit and then I will go over it!

sashadev-sky commented:
@rexagod Ok I rebased it was a tough rebase so let me know if everything checks out with you. I double checked and I think everything works the same but you should confirm that so I can undo it ASAP if not!

rexagod commented:
Okay so for the `_disable` console command I thought that it would be nice to have the mapper toggled back into a small button, because if the user never wanted a mapper in the first place, they would've `keymapper.disable()`ed it in the source code itself. But let me know if you want that changed and I'll get to it!

sashadev-sky commented:
@rexagod yes that sounds great! I just wanted to make sure my rebase didn’t change any of your code functionality

rexagod commented:
@sashadev-sky Just to confirm, is there anything other than the above two commits that's left to be done here (other than the rebase)?

sashadev-sky commented:
@rexagod looking now!

sashadev-sky commented:
@rexagod could you please rebase first because we updated the toolbar to use the newer version of `Leaflet.toolbar` and the old npm package is uninstalled

rexagod commented:
@sashadev-sky I've rebased the branch (which you can have a look over https://github.com/rexagod/Leaflet.DistortableImage/tree/kp2-temp here) but currently I can't see the <i> icons in the keymapper. Is this related to the svg-sprite stuff?

![image](https://user-images.githubusercontent.com/33557095/61010259-29e67580-a393-11e9-8df7-a4a51102e360.png)

rexagod commented:
Actually let me just push it over here.

rexagod commented:
Done. This should make things easier.

sashadev-sky commented:
I'm looking at this rn!

sashadev-sky commented:
@rexagod made some updates to this PR so that it toggles (doesn't close forever). I also refactored the instantiation so that it doesnt have to be done in the html file, it just instantiates a keymapper depending on the `keymapper` option. (still have a little more refactoring to do there).

We're using an SVG icon system now! Check out the wikis! What is your opinion on these icons? I tried using a keyboard with a cross through it as close 
![image](https://user-images.githubusercontent.com/41092741/61247179-8e505e80-a71e-11e9-9c7b-14385b553d32.png)

but this looked a bit misleading, like it was suggesting the user can't use the keybindings if they close the keybinding component. Open to suggestions - we can use any icons from anywhere with svg.

![keymap](https://user-images.githubusercontent.com/41092741/61247428-0880e300-a71f-11e9-9155-c3597af9d39f.gif)

sashadev-sky commented:
@rexagod also could use some advice on the keymapper class! 
1) I still can't figure out a usage for making it a handler instead of just right away control. The only reason this is bothering me is because the [leaflet docs](https://leafletjs.com/examples/extending/extending-3-controls.html) say that handlers should have onAdd and onRemove methods, while controls don't them. And I can't think of what would go in these methods in this scenario. 

2) Why doesn't the click handler work unless we use something like `window.onload`? That is probably a basic JS question but still I wasn't sure

sashadev-sky commented:
Ok so I remembered what we were trying to do was make it on the map and use `enable()` and `disable()`. So I removed all the logic from the overlay for creating it but I am still finding that enable and disable doesn't do anything!

rexagod commented:
@sashadev-sky I've fixed the `disable` issue and added warn logs for multiple declaration as well. Also,

> I still can't figure out a usage for making it a handler instead of just right away control.

It is actually `control` at its heart. I've just extended this from a `handler` **wrapper** for the best of two functionalities.

```js
 _setMapper: function () {
    this._keymapper = L.control({ position: this._position });
```

> Why doesn't the click handler work unless we use something like window.onload?

Probably because the elements aren't rendered before load so there won't be anything to click on.


Let's get this one in! There a lot of PRs that are stalled right now because of this and can easily be merged once this is. What do you think?

jywarren commented:
This is very cool. Great teamwork!

On Sat, Jul 20, 2019, 1:35 PM Pranshu Srivastava <notifications@github.com>
wrote:

> @sashadev-sky <https://github.com/sashadev-sky> I've fixed the disable
> issue and added warn logs for multiple declaration as well. Also,
>
> I still can't figure out a usage for making it a handler instead of just
> right away control.
>
> It is actually control at its heart. I've just extended this from a
> handler *wrapper* for the best of two functionalities.
>
>  _setMapper: function () {
>     this._keymapper = L.control({ position: this._position });
>
> Why doesn't the click handler work unless we use something like
> window.onload?
>
> Probably because the elements aren't rendered before load so there won't
> be anything to click on.
>
> Let's get this one in! There a lot of PRs that are stalled right now
> because of this and can easily be merged once this is. What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/Leaflet.DistortableImage/pull/243?email_source=notifications&email_token=AAAF6JZ2PS6AR54EDC264VDQANEDTA5CNFSM4HKUZDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2NSXOI#issuecomment-513485753>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAF6JZV4GHVSMXDRX6P2LDQANEDTANCNFSM4HKUZDDQ>
> .
>

sashadev-sky commented:
@rexagod Ok awesome! `disable()` is working now, but `enable()` still is not. This is because `L.Handler` instances require `addHooks` and `removeHooks`. So we need to fix that up.

rexagod commented:
Okay so `removeHooks` have been added. But I'm quite a bit unsure as to what to add in `addHooks` because the keymapper is already added to the map once it's initialized, what do you think?

sashadev-sky commented:
@rexagod this is where I was also unsure! Which is why i thought to maybe scrap the `handler` part all together. If we can get it working totally on board with it! I could see perhaps,  in the future, the actual parts of the keymapper could be instances of L.Handler, like Justin Manley does quite cleverly in Leaflet.toolbar for the `action`s. But the control itself I am not sure how to make work like this! I don't know if we need an `enable` and `disable` for it. It could just be a Control and we can add dynamic methods to add it or remove it via `removeLayer`. The user doesn't have to initialize it in the first place if they don't want to anyway

I would be glad to take it a stab at it if your onboard or wait and see if you can figure out this enable thing :)

rexagod commented:
@sashadev-sky Check it out!

sashadev-sky commented:
@rexagod checking it out now!

sashadev-sky commented:
@rexagod ok this is so close! i just got enable and disable working no issues 👍now just have to rebase, cleanup, get your approval, update docs

sashadev-sky commented:
@rexagod this is so close to done!!! So I got rid of the `enable` and `disable` methods because those come with handlers built in. And they trigger `addHooks` / `removeHooks` so I put initialization logic in there. Enable and disable working perfectly now 👏👏

Now I am confused over another issue - why is it impossible to set the position anyway other than `topright`? if try, the button just doesnt show at all

sashadev-sky commented:
@rexagod if you try it out and you're really not sure either I would be ok with merging this anyways and opening a new issue for it! We made a lot of progress here on the keymapper

sashadev-sky commented:
@rexagod Its ready!! Check it out and let me know if you're ready to merge it as well :) We can add some documentation in this PR or a new one - working features currently include 1) enable, 2)disable, 3) positioning 
![keymap](https://user-images.githubusercontent.com/41092741/62043675-ef703b80-b1ce-11e9-94f8-b8854446a08e.gif)

@jywarren see our fancy new keymapper above ^

sashadev-sky commented:
@rexagod can i get some help here pls :) 
Travis build failing - suspicious warnings:
![image](https://user-images.githubusercontent.com/41092741/62094541-824ebb80-b24b-11e9-9a5c-388ff5e35306.png)

Then a few toolbar related tests fail which pass locally. 

Not sure why its failing. Thinking maybe because I added files onto a fork thats not mine? No idea. Could you try please running it for me?
